coil-compose-base/src/androidTest/java/coil/compose/AsyncImagePainterTest.kt

```diff
@@ -612,6 +613,84 @@ class AsyncImagePainterTest {
             .assertIsSimilarTo(R.drawable.sample, threshold = 0.85)
     }
 
+    @Test
+    fun successCallbackIsCalled() {
+        val loadingCount = AtomicInteger()
+        val successCount = AtomicInteger()
+        val errorCount = AtomicInteger()
+
+        composeTestRule.setContent {
+            Image(
+                painter = rememberAsyncImagePainter(
+                    model = ImageRequest.Builder(LocalContext.current)
+                        .data(server.url("/image"))
+                        .size(100, 100)
+                        .build(),
+                    imageLoader = imageLoader,
+                    onLoading = { loadingCount.getAndIncrement() },
+                    onSuccess = { successCount.getAndIncrement() },
+                    onError = { errorCount.getAndIncrement() },
+                ),
+                contentDescription = null
+            )
+        }
+
+        waitForRequestComplete()
+
+        assertEquals(1, loadingCount.get())
+        assertEquals(1, successCount.get())
+        assertEquals(0, errorCount.get())
+    }
+
+    @Test
+    fun errorCallbackIsCalled() {
+        val loadingCount = AtomicInteger()
+        val successCount = AtomicInteger()
+        val errorCount = AtomicInteger()
+
+        composeTestRule.setContent {
+            Image(
+                painter = rememberAsyncImagePainter(
+                    model = ImageRequest.Builder(LocalContext.current)
+                        .data(server.url("/incorrect_path"))
+                        .size(100, 100)
+                        .build(),
+                    imageLoader = imageLoader,
+                    onLoading = { loadingCount.getAndIncrement() },
+                    onSuccess = { successCount.getAndIncrement() },
+                    onError = { errorCount.getAndIncrement() },
+                ),
+                contentDescription = null
+            )
+        }
+
+        waitForRequestComplete()
+
+        assertEquals(1, loadingCount.get())
+        assertEquals(0, successCount.get())
+        assertEquals(1, errorCount.get())
+    }
+
+    @Test
+    fun doesNotRecompose() {
```

Comment:
This test currently fails, but a similar test for `AsyncImage` passes. If you know why, please let me know.

Reply:
Called `AsyncImagePainter.onRemembered` synchronously was the cause.